packages/api/src/external/carequality/document/query-documents.ts

```diff
@@ -0,0 +1,213 @@
+
+  log(`I have ${docsToDownload.length} docs to download (${convertibleDocCount} convertible)`);
+
+  appendDocQueryProgress({
+    patient: { id: patientId, cxId: cxId },
+    downloadProgress: {
+      status: "processing",
+      total: docsToDownload.length,
+    },
+    convertProgress: {
+      status: "processing",
+      total: convertibleDocCount,
+    },
+    requestId,
+    source: MedicalDataSource.COMMONWELL,
+  });
+
+  // TODO - INTRODUCED WHEN IMPLEMENTING CQ DOC RETRIEVAL
+  // downloadDocs(docsToDownload, patientId, cxId, requestId);
+}
+
+function combineDocRefs(docQueryResults: DocumentQueryResult[]): DocumentReference[] {
+  return docQueryResults.reduce((acc: DocumentReference[], curr) => {
+    return [...acc, ...curr.data.documentReference];
+  }, []);
+}
+
+const getNonExistentDocRefs = async (
+  documents: DocumentReference[],
+  patientId: string,
+  cxId: string
+): Promise<DocumentReference[]> => {
+  const [{ existingDocRefs, nonExistingDocRefs }, FHIRDocRefs] = await Promise.all([
+    filterOutExistingDocRefsS3(documents, patientId, cxId),
+    getDocRefsFromFHIR(cxId, patientId),
+  ]);
+
+  const foundOnStorageButNotOnFHIR = existingDocRefs.filter(
+    f => !FHIRDocRefs.find(d => d.id === f.docUniqueId)
+  );
+
+  const docsToDownload = nonExistingDocRefs.concat(foundOnStorageButNotOnFHIR);
+
+  return docsToDownload;
+};
+
+type ExistentialDocRefs = {
+  existingDocRefs: DocumentReference[];
+  nonExistingDocRefs: DocumentReference[];
+};
+
+const filterOutExistingDocRefsS3 = async (
+  documents: DocumentReference[],
+  patientId: string,
+  cxId: string
+): Promise<ExistentialDocRefs> => {
+  const docIdWithExist = await Promise.allSettled(
+    documents.map(async (doc): Promise<{ docId: string; exists: boolean }> => {
+      const fileName = createS3FileName(cxId, patientId, doc.docUniqueId);
+
+      const { exists } = await s3Utils.getFileInfoFromS3(fileName, s3BucketName);
+
+      return {
+        docId: doc.docUniqueId,
+        exists,
+      };
+    })
+  );
+
+  const successfulDocs = docIdWithExist.flatMap(ref =>
+    ref.status === "fulfilled" && ref.value ? ref.value : []
+  );
+
+  const existentialDocRefs = documents.reduce(
+    (acc: ExistentialDocRefs, curr) => {
+      for (const succDoc of successfulDocs) {
+        if (succDoc.docId === curr.docUniqueId) {
+          if (succDoc.exists) {
+            acc.existingDocRefs = [...acc.existingDocRefs, curr];
+          } else {
+            acc.nonExistingDocRefs = [...acc.nonExistingDocRefs, curr];
+          }
+        }
+      }
+      return acc;
+    },
+    {
+      existingDocRefs: [],
+      nonExistingDocRefs: [],
+    }
+  );
+
+  return existentialDocRefs;
+};
+
+const getDocRefsFromFHIR = (cxId: string, patientId: string): Promise<FHIRDocumentReference[]> => {
+  const fhirApi = makeFhirApi(cxId);
+
+  return getAllPages(() =>
+    fhirApi.searchResourcePages("DocumentReference", `patient=${patientId}`)
+  );
+};
```

Comment:
This is not CQ-specific, right? AFAIK we could reuse `external/fhir`'s `getDocuments()`, no?

Could also update CW's to use the same thing, either this function moved to external/fhir or the existing one getDocuments?

Reply:
For now not worrying about refactoring this